File: ToastAlerts/ToastAlertView.m
//
//  ToastView.m
//  ToastAlerts
//
//  Created by Eduardo Irías on 8/7/15.
//  Copyright (c) 2015 Estamp World. All rights reserved.
//

#import "ToastAlertView.h"
#import "UIViewController+ext.h"

@interface ToastAlertView ()

@property UIView *viewToast;
@property UILabel *messageLabel;
@property UIImageView *alertImageView;

@end

@implementation ToastAlertView

@synthesize frameSize;
@synthesize shouldDismissWithTap;
@synthesize shouldDismissWithTime;
@synthesize dismissTime;
@synthesize message;
@synthesize image;
@synthesize images;
@synthesize shouldRepeatImagesAnimation;
@synthesize timeForImagesInAnimation;

#pragma mark - Init

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setVars];
    }
    
    return self;
}
- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setVars];
    }
    
    return self;
}

- (void) setVars {
    dismissTime = 2.0;
    timeForImagesInAnimation = 0.2;
    images = [[NSArray alloc] init];
    shouldRepeatImagesAnimation = true;
    frameSize = 155;

    [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
    [[NSNotificationCenter defaultCenter] addObserver: self selector:   @selector(deviceOrientationDidChange:) name: UIDeviceOrientationDidChangeNotification object: nil];

}

- (void)deviceOrientationDidChange:(NSNotification *)notification {
    //Obtain current device orientation
    [self setFrame:CGRectMake(0, 0, [self screenSize].width, [self screenSize].height) ];
}

#pragma mark - View Cycle

- (void)willMoveToWindow {
    self.backgroundColor = [UIColor clearColor];
}

- (void) didMoveToWindow {
    [self setFrame:CGRectMake(0, 0, [self screenSize].width, [self screenSize].height) ];
    
    self.backgroundColor = [UIColor clearColor];
    
    [self createViewToast];
    [self createMessageLabel];
    [self createalertImageView];
    
    
    if (images.count > 0) {
        _alertImageView.image = images[images.count-1];
        [_alertImageView startAnimating];
    }
    if (shouldDismissWithTap) {
        UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissMessageView:)];
        tapGestureRecognizer.delegate = self;
        [self addGestureRecognizer:tapGestureRecognizer];
    }
    if (shouldDismissWithTime) {
        [NSTimer scheduledTimerWithTimeInterval:dismissTime target:self selector:@selector(dismissMessageView) userInfo:nil repeats:NO];
    }
}

-(void) dealloc{
    [[NSNotificationCenter defaultCenter] removeObserver: self];
    [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];
}

#pragma mark - SubViews

- (void) createViewToast {
    _viewToast = [[UIView alloc] init];
    _viewToast.backgroundColor = [UIColor colorWithRed:222.0/255.0 green:223.0/255.0 blue:227.0/255.0 alpha:1.0];
    _viewToast.layer.cornerRadius = 10;
    _viewToast.layer.masksToBounds = YES;
    [self addSubview:_viewToast];
    
    [_viewToast setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_viewToast
      attribute:NSLayoutAttributeWidth
      relatedBy:NSLayoutRelationEqual
      toItem:nil
      attribute:NSLayoutAttributeWidth
      multiplier:1
      constant:frameSize]
     ];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_viewToast
      attribute:NSLayoutAttributeHeight
      relatedBy:NSLayoutRelationEqual
      toItem:nil
      attribute:NSLayoutAttributeHeight
      multiplier:1
      constant:frameSize]
     ];
    
    
    [self addConstraint:
     [NSLayoutConstraint constraintWithItem:_viewToast
                                  attribute:NSLayoutAttributeCenterX
                                  relatedBy:0
                                     toItem:self
                                  attribute:NSLayoutAttributeCenterX
                                 multiplier:1
                                   constant:0]];
    
    [self addConstraint:
     [NSLayoutConstraint constraintWithItem:_viewToast
                                  attribute:NSLayoutAttributeCenterY
                                  relatedBy:0
                                     toItem:self
                                  attribute:NSLayoutAttributeCenterY
                                 multiplier:1
                                   constant:0]];

}

- (void) createMessageLabel {
    _messageLabel = [[UILabel alloc] init];
    _messageLabel.text = message;
    _messageLabel.font = [UIFont systemFontOfSize:16.0];
    _messageLabel.textColor = [UIColor colorWithRed:90.0/255.0 green:94.0/255.0 blue:97.0/255.0 alpha:1.0];
    [_messageLabel setTextAlignment:NSTextAlignmentCenter];
    [_viewToast addSubview:_messageLabel];
    
    [_messageLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_messageLabel
      attribute:NSLayoutAttributeTrailing
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeTrailing
      multiplier:1
      constant:0]
     ];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_messageLabel
      attribute:NSLayoutAttributeLeading
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeLeading
      multiplier:1
      constant:0]
     ];
    
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_messageLabel
      attribute:NSLayoutAttributeBaseline
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeBaseline
      multiplier:1
      constant:-20]
     ];
    
    
    [_messageLabel addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_messageLabel
      attribute:NSLayoutAttributeHeight
      relatedBy:NSLayoutRelationEqual
      toItem:nil
      attribute:NSLayoutAttributeHeight
      multiplier:1
      constant:32]
     ];
    
    
}

- (void) createalertImageView {
    _alertImageView = [[UIImageView alloc] init];
    _alertImageView.clipsToBounds = true;
    [_alertImageView setContentMode:UIViewContentModeScaleAspectFit];
    
    [_viewToast addSubview:_alertImageView];
    
    if (images.count > 0) {
        _alertImageView.image = images[0];
        _alertImageView.animationImages = images;
        _alertImageView.animationDuration = timeForImagesInAnimation * images.count;
        if (!shouldRepeatImagesAnimation) {
            _alertImageView.animationRepeatCount = 1;
        }
    } else {
        _alertImageView.image = image;
    }
    [_alertImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_alertImageView
      attribute:NSLayoutAttributeTrailing
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeTrailing
      multiplier:1
      constant:-10]
     ];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_alertImageView
      attribute:NSLayoutAttributeLeading
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeLeading
      multiplier:1
      constant:10]
     ];
    
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_alertImageView
      attribute:NSLayoutAttributeTop
      relatedBy:NSLayoutRelationEqual
      toItem:_viewToast
      attribute:NSLayoutAttributeTop
      multiplier:1
      constant:20]
     ];
    
    [_viewToast addConstraint:
     [NSLayoutConstraint
      constraintWithItem:_alertImageView
      attribute:NSLayoutAttributeBaseline
      relatedBy:NSLayoutRelationEqual
      toItem:_messageLabel
      attribute:NSLayoutAttributeTop
      multiplier:1
      constant:-10]
     ];
    
    
    
}

#pragma mark - Dismiss

- (void) dismissMessageView  {
    [self dismiss];
}

- (void) dismissMessageView: (UIGestureRecognizer *) sender {
    [self dismissMessageView];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

#pragma mark - Display

- (void) show {
    UIWindow *frontWindow = [[UIApplication sharedApplication] keyWindow];
    [frontWindow addSubview:self];
}

- (void) dismiss {
    [self removeFromSuperview];
}

- (CGSize)screenSize {
    CGSize screenSize = [UIScreen mainScreen].bounds.size;
    if ((NSFoundationVersionNumber <= NSFoundationVersionNumber_iOS_7_1) && UIInterfaceOrientationIsLandscape([UIApplication sharedApplication].statusBarOrientation)) {
        return CGSizeMake(screenSize.height, screenSize.width);
    }
    return screenSize;
}


@end

File: Sources/Toast Alerts/ToastAlertView.swift
//
//  ToastAlertView.swift
//  Toast Alerts
//
//  Created by Eduardo Irias on 1/31/20.
//  Copyright © 2020 Eduardo Irias. All rights reserved.
//

import UIKit

public class ToastAlertView: UIView {
    
    private let size = CGSize(width: 250, height: 266)
    static let toastTextColor = UIColor(red: 116.0 / 255.0, green: 117.0 / 255.0, blue: 119.0 / 255.0, alpha: 1.0)
    
    // MARK: - Attributes
    
    /**
     Status of the view if it should dismiss with a tap
     
     - Default: true
     */
    open var shouldDismissWithTap: Bool {
        switch dismissType {
        case .tap, .tapAndTime(_):
            return true
        default:
            return false
        }
    }
    
    /**
     Status of the view if it should dismiss with time
     The default value is false
     @see dismissTime
     */
    open var shouldDismissWithTime: Bool {
        switch dismissType {
        case .time(_), .tapAndTime(_):
            return true
        default:
            return false
        }
    }
    
    /**
     The view will dismiss with this time
     The default value is 2.0
     @see shouldDismissWithTime
     */
    open var dismissTime: TimeInterval {
        switch dismissType {
        case .time(let time):
            return time
        case .tapAndTime(let time):
            return time
        default:
            return 0.0
        }
    }
    
    var dismissType = DissmisType.tapAndTime(time: 2.0)

    /**
     The message that should be display
     */
    public var message: String? {
        didSet {
            if let messageLabel = messageLabel {
                messageLabel.text = message
            } else {
                addMessageLabel()
            }
        }
    }
    
    /**
     The image that should be display
     */
    public var image: UIImage? {
        didSet {
            if let alertImageView = alertImageView {
                alertImageView.image = image
            } else {
                addImageView()
            }
        }
    }
    
    private var messageLabel: UILabel?
    private var alertImageView: UIImageView?
    
    // MARK: - Init
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        layoutDisplay()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    /// A new ToastAlertView with message and image
    /// - Parameters:
    ///   - message: The message that will be displayed
    ///   - image: The image that will be displayed
    ///   - hideType: Type of dismissal actions
    public init(message: String, image: UIImage, hideWithType hideType: DissmisType = .tapAndTime()) {
        super.init(frame: CGRect.zero)
        
        layoutDisplay()
        addMessageLabel()
        addImageView()
        
        self.message = message
        self.image = image
        
        messageLabel?.text = message
        alertImageView?.image = image
        
        dismissType = hideType
    }
    
    // MARK: - UI
    
    private func addMessageLabel() {

        messageLabel = UILabel()
        
        guard let messageLabel = messageLabel else { return }
        
        messageLabel.numberOfLines = 0
        messageLabel.text = message
        messageLabel.font = UIFont.boldSystemFont(ofSize: 20)
        messageLabel.textColor = ToastAlertView.toastTextColor
        messageLabel.textAlignment = .center
        
        self.addSubview(messageLabel)
        
        messageLabel.translatesAutoresizingMaskIntoConstraints = false
        
        self.addConstraint(NSLayoutConstraint(item: messageLabel, attribute: .trailing, relatedBy: .equal, toItem: self, attribute: .trailing, multiplier: 1, constant: 0))

        self.addConstraint(NSLayoutConstraint(item: messageLabel, attribute: .leading, relatedBy: .equal, toItem: self, attribute: .leading, multiplier: 1, constant: 0))

        self.addConstraint(NSLayoutConstraint(item: messageLabel, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1, constant: -32))

    }
    
    private func addImageView() {
        alertImageView = UIImageView()
        
        guard let alertImageView = alertImageView else { return }
        
        let padding: CGFloat = 80
        
        alertImageView.clipsToBounds = true
        alertImageView.contentMode = .scaleAspectFit
        alertImageView.image = image
        alertImageView.tintColor = ToastAlertView.toastTextColor
        self.addSubview(alertImageView)
        
        alertImageView.translatesAutoresizingMaskIntoConstraints = false
        
        self.addConstraint(NSLayoutConstraint(item: alertImageView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1.0, constant: 68))
        
        self.addConstraint(NSLayoutConstraint(item: alertImageView, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1.0, constant: 100))

        self.addConstraint(NSLayoutConstraint(item: alertImageView, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0))
        
        self.addConstraint(NSLayoutConstraint(item: alertImageView, attribute: .top, relatedBy: .equal, toItem: self, attribute: .topMargin, multiplier: 1, constant: padding))
        
        //self.addConstraint(NSLayoutConstraint(item: alertImageView, attribute: .bottom, relatedBy: .greaterThanOrEqual, toItem: messageLabel, attribute: .top, multiplier: 1, constant: 0))
    }
    
    private func layoutDisplay() {
        self.isHidden = true
        reCenter()
        
        self.backgroundColor = UIColor.clear
        self.layer.cornerRadius = 8
        self.layer.masksToBounds = true
        NotificationCenter.default.addObserver(self, selector: #selector(reCenter), name: UIDevice.orientationDidChangeNotification, object: nil)
        
        let blurEffect = UIBlurEffect(style: UIBlurEffect.Style.prominent)
        
        let blurEffectView = UIVisualEffectView(effect: blurEffect)
        blurEffectView.frame = self.bounds
        blurEffectView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        blurEffectView.alpha = 0.9
        self.addSubview(blurEffectView)
    }
    
    // MARK: - Display
    
    @objc private func reCenter() {
        let screenSize = UIScreen.main.bounds.size
        let x = (screenSize.width / 2.0) - size.width / 2
        let y = (screenSize.height / 2.0)  - size.height / 2
        frame = CGRect(x: x, y: y, width: size.width, height: size.height)
    }
    
    public func show() {
        let mainWindow = UIApplication.shared.windows.first
        mainWindow?.addSubview(self)
        self.isHidden = false
        
        if shouldDismissWithTime {
            Timer.scheduledTimer(withTimeInterval: dismissTime, repeats: false) { (timer) in
                self.hide()
            }
        }
    }
    
    public func hide() {
        NotificationCenter.default.removeObserver(self, name: UIDevice.orientationDidChangeNotification, object: nil)
        self.isHidden = true
        self.removeFromSuperview()
    }
}

extension ToastAlertView {
    public enum DissmisType {
        case none
        case tap
        case time(time: TimeInterval = 2.0)
        case tapAndTime(time: TimeInterval = 2.0)
    }
}
